File: apps/demo-be-fsharp-giraffe/tests/DemoBeFsgi.Tests/Integration/Steps/ExpenseSteps.fs
module DemoBeFsgi.Tests.Integration.Steps.ExpenseSteps

open System
open System.Text.Json
open TickSpec
open Xunit
open DemoBeFsgi.Tests.State
open DemoBeFsgi.Tests.DirectServices
open DemoBeFsgi.Tests.Integration.Steps.CommonSteps

/// Parse a JSON expense body and call createExpense directly.
let internal createExpenseFromBody (state: StepState) (bodyStr: string) : int * string * string option =
    try
        let doc = JsonDocument.Parse(bodyStr)
        let r = doc.RootElement

        let str (key: string) =
            match r.TryGetProperty(key) with
            | true, el -> el.GetString()
            | _ -> null

        let floatOpt (key: string) =
            match r.TryGetProperty(key) with
            | true, el when el.ValueKind = JsonValueKind.Number -> Some(el.GetDouble())
            | _ -> None

        let unitOpt =
            match r.TryGetProperty("unit") with
            | true, el when el.ValueKind = JsonValueKind.String -> Some(el.GetString())
            | _ -> None

        let status, responseBody =
            createExpense
                state.Db
                state.AccessToken
                (str "amount")
                (str "currency")
                (str "category")
                (str "description")
                (str "date")
                (str "type")
                (floatOpt "quantity")
                unitOpt
            |> Async.RunSynchronously

        let expenseId = getStringProp responseBody "id"
        status, responseBody, expenseId
    with ex ->
        400, $"""{{ "error": "Bad Request", "message": "{ex.Message}" }}""", None

[<Given>]
let ``alice has created an entry with body (.+)`` (bodyStr: string) (state: StepState) =
    let _status, _body, expenseId = createExpenseFromBody state bodyStr
    { state with ExpenseId = expenseId }

[<Given>]
let ``alice has created an expense with body (.+)`` (bodyStr: string) (state: StepState) =
    let _status, _body, expenseId = createExpenseFromBody state bodyStr
    { state with ExpenseId = expenseId }

[<Given>]
let ``alice has created 3 entries`` (state: StepState) =
    createExpenseFromBody
        state
        """{ "amount": "10.00", "currency": "USD", "category": "food", "description": "Entry 1", "date": "2025-01-01", "type": "expense" }"""
    |> ignore

    createExpenseFromBody
        state
        """{ "amount": "20.00", "currency": "USD", "category": "food", "description": "Entry 2", "date": "2025-01-02", "type": "expense" }"""
    |> ignore

    createExpenseFromBody
        state
        """{ "amount": "30.00", "currency": "USD", "category": "food", "description": "Entry 3", "date": "2025-01-03", "type": "expense" }"""
    |> ignore

    state

[<When>]
let ``alice sends POST /api/v1/expenses with body (.+)`` (bodyStr: string) (state: StepState) =
    let status, responseBody, _expenseId = createExpenseFromBody state bodyStr

    { state with
        Response = Some { Status = status; Body = responseBody }
        ResponseBody = Some responseBody }

[<When>]
let ``alice sends GET /api/v1/expenses/\{expenseId\}`` (state: StepState) =
    let expenseId =
        state.ExpenseId
        |> Option.bind (fun s ->
            try
                Some(Guid.Parse(s))
            with _ ->
                None)

    match expenseId with
    | Some id ->
        let status, body = getExpenseById state.Db state.AccessToken id |> Async.RunSynchronously

        { state with
            Response = Some { Status = status; Body = body }
            ResponseBody = Some body }
    | None ->
        { state with
            Response = Some { Status = 404; Body = """{"error":"Not Found"}""" }
            ResponseBody = Some """{"error":"Not Found"}""" }

[<When>]
let ``alice sends GET /api/v1/expenses`` (state: StepState) =
    let status, body = listExpenses state.Db state.AccessToken 1 20 |> Async.RunSynchronously

    { state with
        Response = Some { Status = status; Body = body }
        ResponseBody = Some body }

[<When>]
let ``alice sends PUT /api/v1/expenses/\{expenseId\} with body (.+)`` (bodyStr: string) (state: StepState) =
    let expenseId =
        state.ExpenseId
        |> Option.bind (fun s ->
            try
                Some(Guid.Parse(s))
            with _ ->
                None)

    match expenseId with
    | Some id ->
        let status, responseBody =
            try
                let doc = JsonDocument.Parse(bodyStr)
                let r = doc.RootElement

                let str (key: string) =
                    match r.TryGetProperty(key) with
                    | true, el -> el.GetString()
                    | _ -> null

                updateExpense
                    state.Db
                    state.AccessToken
                    id
                    (str "amount")
                    (str "currency")
                    (str "category")
                    (str "description")
                    (str "date")
                    (str "type")
                |> Async.RunSynchronously
            with ex ->
                400, $"""{{ "error": "Bad Request", "message": "{ex.Message}" }}"""

        { state with
            Response = Some { Status = status; Body = responseBody }
            ResponseBody = Some responseBody }
    | None ->
        { state with
            Response = Some { Status = 404; Body = """{"error":"Not Found"}""" }
            ResponseBody = Some """{"error":"Not Found"}""" }

[<When>]
let ``alice sends DELETE /api/v1/expenses/\{expenseId\}`` (state: StepState) =
    let expenseId =
        state.ExpenseId
        |> Option.bind (fun s ->
            try
                Some(Guid.Parse(s))
            with _ ->
                None)

    match expenseId with
    | Some id ->
        let status, body = deleteExpense state.Db state.AccessToken id |> Async.RunSynchronously

        { state with
            Response = Some { Status = status; Body = body }
            ResponseBody = Some body }
    | None ->
        { state with
            Response = Some { Status = 404; Body = """{"error":"Not Found"}""" }
            ResponseBody = Some """{"error":"Not Found"}""" }

[<When>]
let ``alice sends GET /api/v1/expenses/summary`` (state: StepState) =
    let status, body = expenseSummary state.Db state.AccessToken |> Async.RunSynchronously

    { state with
        Response = Some { Status = status; Body = body }
        ResponseBody = Some body }

[<Then>]
let ``the response body should contain "(.+)" total equal to "(.+)"``
    (currency: string)
    (expected: string)
    (state: StepState)
    =
    let body = state.ResponseBody.Value

    try
        let doc = JsonDocument.Parse(body)
        let el = doc.RootElement.GetProperty(currency)
        let actual = el.GetString()
        Assert.Equal(expected, actual)
    with ex ->
        Assert.Fail($"Could not find currency '{currency}' in response: {body}. Error: {ex.Message}")

    state

[<Then>]
let ``the response body should contain "quantity" equal to (.+)`` (expected: string) (state: StepState) =
    let body = state.ResponseBody.Value

    try
        let doc = JsonDocument.Parse(body)
        let el = doc.RootElement.GetProperty("quantity")

        match el.ValueKind with
        | JsonValueKind.Number ->
            let actual = el.GetDouble()
            let expectedVal = float expected
            Assert.Equal(expectedVal, actual)
        | JsonValueKind.Object ->
            // Optional field wrapped in Some
            ()
        | _ -> Assert.Fail($"Expected numeric 'quantity' field in: {body}")
    with ex ->
        Assert.Fail($"Could not find 'quantity' in response: {body}. Error: {ex.Message}")

    state

// Note: "unit" field assertion is handled by CommonSteps ``the response body should contain "(.+)" equal to "(.+)"``

File: apps/demo-be-kotlin-ktor/src/test/kotlin/com/demobektkt/integration/steps/ExpenseSteps.kt
package com.demobektkt.integration.steps

import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.junit.jupiter.api.Assertions.assertTrue

class ExpenseSteps {

    private fun createExpense(username: String, body: Map<String, String?>): String {
        val token = TestWorld.accessTokens[username] ?: error("$username has no access token")
        val amount = body["amount"] ?: error("amount missing")
        val currency = body["currency"] ?: error("currency missing")
        val category = body["category"] ?: error("category missing")
        val description = body["description"] ?: error("description missing")
        val date = body["date"] ?: error("date missing")
        val type = body["type"] ?: error("type missing")
        val quantity = body["quantity"]?.toDoubleOrNull()
        val unit = body["unit"]
        val (status, respBody) =
            ServiceDispatcher.createExpense(
                token,
                amount,
                currency,
                category,
                description,
                date,
                type,
                quantity,
                unit,
            )
        assertTrue(status == 201, "Expected 201 creating expense, got $status. Body: $respBody")
        return JsonHelper.getString(respBody, "id") ?: error("No id in expense response: $respBody")
    }

    @Suppress("NestedBlockDepth")
    private fun loginAndEnsureToken(username: String, passwords: List<String>) {
        if (!TestWorld.accessTokens.containsKey(username)) {
            for (pwd in passwords) {
                val (status, body) = ServiceDispatcher.login(username, pwd)
                if (status == 200) {
                    JsonHelper.getString(body, "access_token")?.let {
                        TestWorld.accessTokens[username] = it
                    }
                    JsonHelper.getString(body, "refresh_token")?.let {
                        TestWorld.refreshTokens[username] = it
                    }
                    break
                }
            }
        }
    }

    @Given(
        "alice has created an entry with body \\{ {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string} \\}"
    )
    fun aliceHasCreatedEntryWithBody(
        k1: String,
        v1: String,
        k2: String,
        v2: String,
        k3: String,
        v3: String,
        k4: String,
        v4: String,
        k5: String,
        v5: String,
        k6: String,
        v6: String,
    ) {
        val body = mapOf(k1 to v1, k2 to v2, k3 to v3, k4 to v4, k5 to v5, k6 to v6)
        val id = createExpense("alice", body)
        TestWorld.expenseIds["alice:last"] = id
    }

    @Given(
        "alice has created an expense with body \\{ {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string} \\}"
    )
    fun aliceHasCreatedExpenseWithSixFields(
        k1: String,
        v1: String,
        k2: String,
        v2: String,
        k3: String,
        v3: String,
        k4: String,
        v4: String,
        k5: String,
        v5: String,
        k6: String,
        v6: String,
    ) {
        val body = mapOf(k1 to v1, k2 to v2, k3 to v3, k4 to v4, k5 to v5, k6 to v6)
        val id = createExpense("alice", body)
        TestWorld.expenseIds["alice:last"] = id
    }

    @Given(
        "^alice has created an entry with body \\{ \"amount\": \"([^\"]+)\", \"currency\": \"([^\"]+)\", \"category\": \"([^\"]+)\", \"description\": \"([^\"]+)\", \"date\": \"([^\"]+)\", \"type\": \"([^\"]+)\", \"quantity\": ([0-9.]+), \"unit\": \"([^\"]+)\" \\}$"
    )
    fun aliceHasCreatedEntryWithQuantityUnit(
        amount: String,
        currency: String,
        category: String,
        description: String,
        date: String,
        type: String,
        quantity: String,
        unit: String,
    ) {
        val body =
            mapOf(
                "amount" to amount,
                "currency" to currency,
                "category" to category,
                "description" to description,
                "date" to date,
                "type" to type,
                "quantity" to quantity,
                "unit" to unit,
            )
        val id = createExpense("alice", body)
        TestWorld.expenseIds["alice:last"] = id
    }

    @Given(
        "^alice has created an expense with body \\{ \"amount\": \"([^\"]+)\", \"currency\": \"([^\"]+)\", \"category\": \"([^\"]+)\", \"description\": \"([^\"]+)\", \"date\": \"([^\"]+)\", \"type\": \"([^\"]+)\", \"quantity\": ([0-9.]+), \"unit\": \"([^\"]+)\" \\}$"
    )
    fun aliceHasCreatedExpenseWithQuantityUnit(
        amount: String,
        currency: String,
        category: String,
        description: String,
        date: String,
        type: String,
        quantity: String,
        unit: String,
    ) {
        val body =
            mapOf(
                "amount" to amount,
                "currency" to currency,
                "category" to category,
                "description" to description,
                "date" to date,
                "type" to type,
                "quantity" to quantity,
                "unit" to unit,
            )
        val id = createExpense("alice", body)
        TestWorld.expenseIds["alice:last"] = id
    }

    @Given("alice has created 3 entries")
    fun aliceHasCreated3Entries() {
        val entries =
            listOf(
                mapOf(
                    "amount" to "10.00",
                    "currency" to "USD",
                    "category" to "food",
                    "description" to "Entry1",
                    "date" to "2025-01-01",
                    "type" to "expense",
                ),
                mapOf(
                    "amount" to "20.00",
                    "currency" to "USD",
                    "category" to "food",
                    "description" to "Entry2",
                    "date" to "2025-01-02",
                    "type" to "expense",
                ),
                mapOf(
                    "amount" to "30.00",
                    "currency" to "USD",
                    "category" to "food",
                    "description" to "Entry3",
                    "date" to "2025-01-03",
                    "type" to "expense",
                ),
            )
        entries.forEach { createExpense("alice", it) }
    }

    @When("alice sends GET \\/api\\/v1\\/expenses\\/\\{expenseId\\}")
    fun aliceSendsGetExpenseById() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
        val (status, body) = ServiceDispatcher.getExpenseById(token, expenseId)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When("alice sends GET \\/api\\/v1\\/expenses")
    fun aliceSendsGetExpenses() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val (status, body) = ServiceDispatcher.listExpenses(token)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When("alice sends GET \\/api\\/v1\\/expenses\\/summary")
    fun aliceSendsGetExpensesSummary() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val (status, body) = ServiceDispatcher.expenseSummary(token)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When(
        "^alice sends PUT /api/v1/expenses/\\{expenseId\\} with body \\{ \"amount\": \"([^\"]+)\", \"currency\": \"([^\"]+)\", \"category\": \"([^\"]+)\", \"description\": \"([^\"]+)\", \"date\": \"([^\"]+)\", \"type\": \"([^\"]+)\" \\}$"
    )
    fun aliceSendsPutExpenseWithBody(
        amount: String,
        currency: String,
        category: String,
        description: String,
        date: String,
        type: String,
    ) {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
        val (status, body) =
            ServiceDispatcher.updateExpense(
                token,
                expenseId,
                amount,
                currency,
                category,
                description,
                date,
                type,
            )
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When("alice sends DELETE \\/api\\/v1\\/expenses\\/\\{expenseId\\}")
    fun aliceSendsDeleteExpense() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
        val (status, body) = ServiceDispatcher.deleteExpense(token, expenseId)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When(
        "^alice sends POST /api/v1/expenses with body \\{ \"amount\": \"([^\"]+)\", \"currency\": \"([^\"]+)\", \"category\": \"([^\"]+)\", \"description\": \"([^\"]+)\", \"date\": \"([^\"]+)\", \"type\": \"([^\"]+)\" \\}$"
    )
    fun aliceSendsPostExpensesWithBody(
        amount: String,
        currency: String,
        category: String,
        description: String,
        date: String,
        type: String,
    ) {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val (status, body) =
            ServiceDispatcher.createExpense(token, amount, currency, category, description, date, type)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
        if (status == 201) {
            JsonHelper.getString(body, "id")?.let { TestWorld.expenseIds["alice:last"] = it }
        }
    }

    @When(
        "^alice sends POST /api/v1/expenses with body \\{ \"amount\": \"([^\"]+)\", \"currency\": \"([^\"]+)\", \"category\": \"([^\"]+)\", \"description\": \"([^\"]+)\", \"date\": \"([^\"]+)\", \"type\": \"([^\"]+)\", \"quantity\": ([0-9.]+), \"unit\": \"([^\"]+)\" \\}$"
    )
    fun aliceSendsPostExpensesWithBodyAndUnit(
        amount: String,
        currency: String,
        category: String,
        description: String,
        date: String,
        type: String,
        quantity: String,
        unit: String,
    ) {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val (status, body) =
            ServiceDispatcher.createExpense(
                token,
                amount,
                currency,
                category,
                description,
                date,
                type,
                quantity.toDoubleOrNull(),
                unit,
            )
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
        if (status == 201) {
            JsonHelper.getString(body, "id")?.let { TestWorld.expenseIds["alice:last"] = it }
        }
    }

    @Then("the response body should contain {string} total equal to {string}")
    fun theResponseBodyShouldContainCurrencyTotalEqualTo(currency: String, total: String) {
        val body = TestWorld.lastResponseBody
        assertTrue(
            body.contains(currency) && body.contains(total),
            "Expected currency '$currency' with total '$total' in: $body",
        )
    }

    @Then("^the response body should contain \"([^\"]+)\" equal to ([0-9]+(?:\\.[0-9]+)?)$")
    fun theResponseBodyShouldContainFieldEqualToNumber(field: String, expected: String) {
        val actual = JsonHelper.getNumber(TestWorld.lastResponseBody, field)
        val expectedDouble = expected.toDouble()
        val actualDouble = actual?.toDoubleOrNull()
        assertTrue(
            actualDouble != null && actualDouble == expectedDouble,
            "Field '$field' expected $expected but got $actual in: ${TestWorld.lastResponseBody}",
        )
    }
}
